File: src/common/Helpers/ConsoleHelpers.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Runtime.CompilerServices;

public class ConsoleHelpers
{
    public static void ConfigureDebug(bool debug)
    {
        _debug = _debug || debug;
        WriteDebugLine($"Debug: {_debug}");
    }

    public static void Configure(bool debug, bool verbose, bool quiet)
    {
        Console.OutputEncoding = Encoding.UTF8;

        _debug = debug;
        _verbose = verbose;
        _quiet = quiet;

        WriteDebugLine($"Debug: {_debug}");
        WriteDebugLine($"Verbose: {_verbose}");
        WriteDebugLine($"Quiet: {_quiet}");
    }

    public static bool IsQuiet()
    {
        return _quiet;
    }

    public static bool IsVerbose()
    {
        return _verbose;
    }

    public static bool IsDebug()
    {
        return _debug;
    }

    public static void DisplayStatus(string status)
    {
        if (!_debug && !_verbose) return;
        if (Console.IsOutputRedirected) return;

        lock (_printLock)
        {
            DisplayStatusErase();
            Console.Write("\r" + status);
            _cchLastStatus = status.Length;
            if (_debug) Thread.Sleep(1);
        }
    }

    public static void DisplayStatusErase()
    {
        if (!_debug && !_verbose) return;
        if (_cchLastStatus <= 0) return;

        lock (_printLock)
        {
            var eraseLastStatus = "\r" + new string(' ', _cchLastStatus) + "\r";
            Console.Write(eraseLastStatus);
            _cchLastStatus = 0;
        }
    }

    public static void Write(string message, ConsoleColor? foregroundColor = null, bool overrideQuiet = false)
    {
        Write(message, foregroundColor, null, overrideQuiet);
    }

    public static void Write(string message, ConsoleColor? foregroundColor, ConsoleColor? backgroundColor, bool overrideQuiet = false)
    {
        if (_quiet && !overrideQuiet) return;

        lock (_printLock)
        {
            WriteWithColorWithoutScrollSmear(message, foregroundColor, backgroundColor);
        }
    }

    public static void WriteLine(string message = "", ConsoleColor? color = null, bool overrideQuiet = false)
    {
        WriteLine(message, color, null, overrideQuiet);
    }

    public static void WriteLine(string message, ConsoleColor? foregroundColor, ConsoleColor? backgroundColor, bool overrideQuiet = false)
    {
        if (_quiet && !overrideQuiet) return;
        Write(message + '\n', foregroundColor, backgroundColor, overrideQuiet);
    }

    public static void WriteLineIfNotEmpty(string message)
    {
        if (string.IsNullOrEmpty(message)) return;
        WriteLine(message);
    }

    public static void WriteWarning(string message, [CallerFilePath] string callerFilePath = "", [CallerLineNumber] int callerLineNumber = 0)
    {
        Write(message, ConsoleColor.Black, ConsoleColor.Yellow, overrideQuiet: true);
        Logger.LogMessage(LogLevel.Warning, "WARNING:", callerFilePath, callerLineNumber, message);
    }

    public static void WriteWarningLine(string message, [CallerFilePath] string callerFilePath = "", [CallerLineNumber] int callerLineNumber = 0)
    {
        WriteLine(message, ConsoleColor.Black, ConsoleColor.Yellow, overrideQuiet: true);
        Logger.LogMessage(LogLevel.Warning, "WARNING:", callerFilePath, callerLineNumber, message);
    }
    
    public static void WriteError(string message, [CallerFilePath] string callerFilePath = "", [CallerLineNumber] int callerLineNumber = 0)
    {
        Write(message, ConsoleColor.White, ConsoleColor.Red, overrideQuiet: true);
        Logger.LogMessage(LogLevel.Error, "ERROR:", callerFilePath, callerLineNumber, message);
    }

    public static void WriteErrorLine(string message, [CallerFilePath] string callerFilePath = "", [CallerLineNumber] int callerLineNumber = 0)
    {
        WriteLine(message, ConsoleColor.White, ConsoleColor.Red, overrideQuiet: true);
        Logger.LogMessage(LogLevel.Error, "ERROR:", callerFilePath, callerLineNumber, message);
    }

    public static void WriteDebug(string message, [CallerFilePath] string callerFilePath = "", [CallerLineNumber] int callerLineNumber = 0)
    {
        Logger.LogMessage(LogLevel.Verbose, "VERBOSE:", callerFilePath, callerLineNumber, message);
        if (_debug)
        {
            Write(message, ConsoleColor.Cyan);
        }
    }

    public static void WriteDebugLine(string message = "", [CallerFilePath] string callerFilePath = "", [CallerLineNumber] int callerLineNumber = 0)
    {
        Logger.LogMessage(LogLevel.Verbose, "VERBOSE:", callerFilePath, callerLineNumber, message);
        if (_debug)
        {
            WriteLine(message, ConsoleColor.Cyan);
        }
    }

    public static void WriteDebugHexDump(string message, string? title = null)
    {
        var noMessage = string.IsNullOrEmpty(message);
        if (noMessage)
        {
            WriteDebugLine($"{title}\n  0000: (empty)");
            return;
        }

        var i = 0;
        foreach (var ch in message)
        {
            if (i % 16 == 0)
            {
                title = title!.Replace("\r", "\\r").Replace("\n", "\\n");
                WriteDebug(i == 0 ? $"{title}\n" : "\n");
                WriteDebug(string.Format("  {0:x4}: ", i));
            }

            WriteDebug(string.Format("{0:X2} ", (int)ch));
            i++;
        }
        WriteDebugLine();
    }


    public static bool IsStandardInputReference(string fileName)
    {
        return fileName == "-" || fileName == "stdin" || fileName == "STDIN" || fileName == "STDIN:";
    }

    public static IEnumerable<string> GetAllLinesFromStdin()
    {
        return _allLinesFromStdin == null
            ? ReadAllLinesFromStdin()
            : _allLinesFromStdin;
    }

    public static void SetForegroundColor(ConsoleColor color)
    {
        Console.ForegroundColor = ColorHelpers.MapColor(color);
    }

    public static ConsoleKeyInfo? ReadKey(bool intercept = false)
    {
        if (Console.IsInputRedirected)
        {
            var line = Console.ReadLine()?.TrimEnd();
            if (line == null) return null;

            var treatAsEnter = line.Length == 0 || line == "\n" || line == "\r" || line == "\r\n";
            if (treatAsEnter) return new ConsoleKeyInfo('\n', ConsoleKey.Enter, false, false, false);

            var isAlphaNumeric = char.IsLetterOrDigit(line.ElementAt(0)) || char.IsPunctuation(line.ElementAt(0)) || char.IsSymbol(line.ElementAt(0));
            if (isAlphaNumeric) return new ConsoleKeyInfo(line.ElementAt(0), (ConsoleKey)line.ElementAt(0), false, false, false);

            return line.ElementAt(0) switch
            {
                '\t' => new ConsoleKeyInfo('\t', ConsoleKey.Tab, false, false, false),
                ' ' => new ConsoleKeyInfo(' ', ConsoleKey.Spacebar, false, false, false),
                _ => null
            };
        }

        return Console.ReadKey(intercept);
    }


    /// <summary>
    /// Logs an exception with detailed information to both the console and logger system.
    /// </summary>
    /// <param name="ex">The exception to log</param>
    /// <param name="contextMessage">Optional context message to prefix the exception</param>
    /// <param name="showToUser">Whether to show the message to the user on console</param>
    /// <param name="filePath">Source file where the exception was caught</param>
    /// <param name="lineNumber">Line number where the exception was caught</param>
    public static void LogException(Exception ex, string contextMessage = "", bool showToUser = true, [CallerFilePath] string filePath = "", [CallerLineNumber] int lineNumber = 0)
    {
        var message = string.IsNullOrEmpty(contextMessage) 
            ? $"Exception: {ex.Message}" 
            : $"{contextMessage}: {ex.Message}";
    
        // Show in console if requested
        if (showToUser)
        {
            WriteErrorLine(message, filePath, lineNumber);
        }
    
        // Always log to persistent storage with stack trace
        Logger.LogMessage(LogLevel.Error, "ERROR:", filePath, lineNumber, $"{message}\n{ex.StackTrace}");
    
        // Log inner exceptions too
        var inner = ex.InnerException;
        int depth = 0;
        while (inner != null && depth < 5)
        {
            Logger.LogMessage(LogLevel.Error, "ERROR:", filePath, lineNumber, $"Inner exception ({depth}): {inner.Message}\n{inner.StackTrace}");
            inner = inner.InnerException;
            depth++;
        }
    }

    /// <summary>
    /// Displays a user function call in the standard format: "user-function: name => result"
    /// </summary>
    /// <param name="functionName">The name of the function being called</param>
    /// <param name="result">The result of the function call, or null if still executing</param>
    public static void DisplayUserFunctionCall(string functionName, string? result)
    {
        Write($"\ruser-function: {functionName} => ", ConsoleColor.DarkGray);

        if (result == null) Write("...", ConsoleColor.DarkGray);
        if (result != null)
        {
            WriteLine(result, ConsoleColor.DarkGray);
            Write("\rUser: ", ConsoleColor.Green);
            SetForegroundColor(ConsoleColor.White);
        }
    }

    /// <summary>
    /// Displays an assistant function call in the standard format: "assistant-function: name args => result"
    /// </summary>
    /// <param name="name">The name of the function being called</param>
    /// <param name="args">The arguments passed to the function</param>
    /// <param name="result">The result of the function call, or null if still executing</param>
    public static void DisplayAssistantFunctionCall(string name, string args, object? result)
    {
        Write($"\rassistant-function: {name} {args} => ", ConsoleColor.DarkGray);

        if (result == null)
        {
            Write("...", ConsoleColor.DarkGray);
        }
        else
        {
            var text = result as string ?? "non-string result";
            WriteLine(text, ConsoleColor.DarkGray);
        }
    }

    private static List<string> ReadAllLinesFromStdin()
    {
        _allLinesFromStdin = new();
        while (true)
        {
            var line = Console.ReadLine();
            if (line == null) break;

            _allLinesFromStdin.Add(line);
        }

        return _allLinesFromStdin;
    }

    private static void WriteWithColorWithoutScrollSmear(string message, ConsoleColor? foregroundColor, ConsoleColor? backgroundColor)
    {
        var lines = message
            .Split(new[] { '\n' }, StringSplitOptions.None)
            .Select(line => line.TrimEnd('\r'))
            .ToArray();
        for (var i = 0; i < lines.Length; i++)
        {
            if (i > 0) Console.WriteLine();
            WriteWithColor(lines[i], foregroundColor, backgroundColor);
        }
    }

    private static void WriteWithColor(string message, ConsoleColor? foregroundColor, ConsoleColor? backgroundColor)
    {
        var prevForegroundColor = Console.ForegroundColor;
        if (foregroundColor != null) Console.ForegroundColor = ColorHelpers.MapColor((ConsoleColor)foregroundColor);

        var prevBackgroundColor = Console.BackgroundColor;
        if (backgroundColor != null) Console.BackgroundColor = ColorHelpers.MapColor((ConsoleColor)backgroundColor);

        Console.Write(message);
        if (foregroundColor != null) Console.ForegroundColor = prevForegroundColor;
        if (backgroundColor != null) Console.BackgroundColor = prevBackgroundColor;
    }
   
    private static bool _debug = false;
    private static bool _verbose = false;
    private static bool _quiet = false;

    private static object _printLock = new();
    private static int _cchLastStatus = 0;

    private static List<string>? _allLinesFromStdin;
}

File: src/commonjs/Helpers/ConsoleHelpers.ts
import { ColorHelpers } from './ColorHelpers';
import { ConsoleColor } from './Colors';

/**
 * Console I/O operations helper class
 */
export class ConsoleHelpers {
    public static configureDebug(debug: boolean): void
    {
        ConsoleHelpers._debug = ConsoleHelpers._debug || debug;
        ConsoleHelpers.writeDebugLine(`Debug: ${ConsoleHelpers._debug}`);
    }

    public static configure(debug: boolean, verbose: boolean, quiet: boolean): void
    {
        // Node.js uses UTF-8 by default
        process.stdout.setEncoding('utf8');

        ConsoleHelpers._debug = debug;
        ConsoleHelpers._verbose = verbose;
        ConsoleHelpers._quiet = quiet;

        ConsoleHelpers.writeDebugLine(`Debug: ${ConsoleHelpers._debug}`);
        ConsoleHelpers.writeDebugLine(`Verbose: ${ConsoleHelpers._verbose}`);
        ConsoleHelpers.writeDebugLine(`Quiet: ${ConsoleHelpers._quiet}`);
    }

    public static isQuiet(): boolean
    {
        return ConsoleHelpers._quiet;
    }

    public static isVerbose(): boolean
    {
        return ConsoleHelpers._verbose;
    }

    public static isDebug(): boolean
    {
        return ConsoleHelpers._debug;
    }

    public static displayStatus(status: string): void
    {
        if (!ConsoleHelpers._debug && !ConsoleHelpers._verbose) return;
        if (process.stdout.isTTY === false) return;

        ConsoleHelpers.displayStatusErase();
        process.stdout.write('\r' + status);
        ConsoleHelpers._cchLastStatus = status.length;
        if (ConsoleHelpers._debug) {
            // Small delay for debug mode
            setTimeout(() => {}, 1);
        }
    }

    public static displayStatusErase(): void
    {
        if (!ConsoleHelpers._debug && !ConsoleHelpers._verbose) return;
        if (ConsoleHelpers._cchLastStatus <= 0) return;

        const eraseLastStatus = '\r' + ' '.repeat(ConsoleHelpers._cchLastStatus) + '\r';
        process.stdout.write(eraseLastStatus);
        ConsoleHelpers._cchLastStatus = 0;
    }

    public static write(message: string, foregroundColor?: ConsoleColor | null, overrideQuiet: boolean = false): void
    {
        ConsoleHelpers.write(message, foregroundColor, null, overrideQuiet);
    }

    public static write(message: string, foregroundColor?: ConsoleColor | null, backgroundColor?: ConsoleColor | null, overrideQuiet: boolean = false): void
    {
        if (ConsoleHelpers._quiet && !overrideQuiet) return;

        ConsoleHelpers.writeWithColorWithoutScrollSmear(message, foregroundColor, backgroundColor);
    }

    public static writeLine(message: string = '', color?: ConsoleColor | null, overrideQuiet: boolean = false): void
    {
        ConsoleHelpers.writeLine(message, color, null, overrideQuiet);
    }

    public static writeLine(message: string, foregroundColor?: ConsoleColor | null, backgroundColor?: ConsoleColor | null, overrideQuiet: boolean = false): void
    {
        if (ConsoleHelpers._quiet && !overrideQuiet) return;
        ConsoleHelpers.write(message + '\n', foregroundColor, backgroundColor, overrideQuiet);
    }

    public static writeLineIfNotEmpty(message: string): void
    {
        if (!message) return;
        ConsoleHelpers.writeLine(message);
    }

    public static writeWarning(message: string): void
    {
        ConsoleHelpers.write(message, ConsoleColor.Black, ConsoleColor.Yellow, true);
    }

    public static writeWarningLine(message: string): void
    {
        ConsoleHelpers.writeLine(message, ConsoleColor.Black, ConsoleColor.Yellow, true);
    }
    
    public static writeError(message: string): void
    {
        ConsoleHelpers.write(message, ConsoleColor.White, ConsoleColor.Red, true);
    }

    public static writeErrorLine(message: string): void
    {
        ConsoleHelpers.writeLine(message, ConsoleColor.White, ConsoleColor.Red, true);
    }

    public static writeDebug(message: string): void
    {
        if (!ConsoleHelpers._debug) return;
        ConsoleHelpers.write(message, ConsoleColor.Cyan);
    }

    public static writeDebugLine(message: string = ''): void
    {
        if (!ConsoleHelpers._debug) return;
        ConsoleHelpers.writeLine(message, ConsoleColor.Cyan);
    }

    public static writeDebugHexDump(message: string, title?: string | null): void
    {
        const noMessage = !message;
        if (noMessage) {
            ConsoleHelpers.writeDebugLine(`${title}\n  0000: (empty)`);
            return;
        }

        let i = 0;
        for (const ch of message) {
            if (i % 16 === 0) {
                const safeTitle = title?.replace(/\r/g, '\\r').replace(/\n/g, '\\n');
                ConsoleHelpers.writeDebug(i === 0 ? `${safeTitle}\n` : '\n');
                ConsoleHelpers.writeDebug(`  ${i.toString(16).padStart(4, '0')}: `);
            }

            ConsoleHelpers.writeDebug(`${ch.charCodeAt(0).toString(16).padStart(2, '0').toUpperCase()} `);
            i++;
        }
        ConsoleHelpers.writeDebugLine();
    }


    public static isStandardInputReference(fileName: string): boolean
    {
        return fileName === '-' || fileName === 'stdin' || fileName === 'STDIN' || fileName === 'STDIN:';
    }

    public static getAllLinesFromStdin(): string[]
    {
        return ConsoleHelpers._allLinesFromStdin == null
            ? ConsoleHelpers.readAllLinesFromStdin()
            : ConsoleHelpers._allLinesFromStdin;
    }

    public static readKey(intercept: boolean = false): { key: string, keyCode: number } | null
    {
        if (!process.stdin.isTTY) {
            // Input is redirected, read line
            const readline = require('readline');
            const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
            
            return new Promise((resolve) => {
                rl.question('', (line: string) => {
                    rl.close();
                    
                    if (line == null) {
                        resolve(null);
                        return;
                    }
                    
                    const trimmedLine = line.trimEnd();
                    const treatAsEnter = trimmedLine.length === 0 || trimmedLine === '\n' || trimmedLine === '\r' || trimmedLine === '\r\n';
                    if (treatAsEnter) {
                        resolve({ key: '\n', keyCode: 13 });
                        return;
                    }
                    
                    const firstChar = trimmedLine[0];
                    const isAlphaNumeric = /[a-zA-Z0-9\p{P}\p{S}]/u.test(firstChar);
                    if (isAlphaNumeric) {
                        resolve({ key: firstChar, keyCode: firstChar.charCodeAt(0) });
                        return;
                    }
                    
                    switch (firstChar) {
                        case '\t':
                            resolve({ key: '\t', keyCode: 9 });
                            break;
                        case ' ':
                            resolve({ key: ' ', keyCode: 32 });
                            break;
                        default:
                            resolve(null);
                    }
                });
            }) as any;
        }
        
        // For TTY, we'd need additional setup for raw key reading
        // This is a simplified version - in practice you might want to use a library like 'keypress'
        return null;
    }

    private static readAllLinesFromStdin(): string[]
    {
        ConsoleHelpers._allLinesFromStdin = [];
        
        // In Node.js, we need to read from stdin synchronously
        const fs = require('fs');
        try {
            const input = fs.readFileSync(process.stdin.fd, 'utf8');
            ConsoleHelpers._allLinesFromStdin = input.split(/\r?\n/);
            // Remove last empty line if present
            if (ConsoleHelpers._allLinesFromStdin[ConsoleHelpers._allLinesFromStdin.length - 1] === '') {
                ConsoleHelpers._allLinesFromStdin.pop();
            }
        } catch (error) {
            // If stdin is not available or empty, return empty array
            ConsoleHelpers._allLinesFromStdin = [];
        }

        return ConsoleHelpers._allLinesFromStdin;
    }

    private static writeWithColorWithoutScrollSmear(message: string, foregroundColor?: ConsoleColor | null, backgroundColor?: ConsoleColor | null): void
    {
        const lines = message
            .split('\n')
            .map(line => line.replace(/\r$/, ''));
        
        for (let i = 0; i < lines.length; i++) {
            if (i > 0) process.stdout.write('\n');
            ConsoleHelpers.writeWithColor(lines[i], foregroundColor, backgroundColor);
        }
    }

    private static writeWithColor(message: string, foregroundColor?: ConsoleColor | null, backgroundColor?: ConsoleColor | null): void
    {
        // In Node.js, we use ANSI escape codes for colors
        let output = '';
        
        if (foregroundColor != null) {
            const mappedColor = ColorHelpers.mapColor(foregroundColor);
            output += ConsoleHelpers.getAnsiColorCode(mappedColor, false);
        }
        
        if (backgroundColor != null) {
            const mappedColor = ColorHelpers.mapColor(backgroundColor);
            output += ConsoleHelpers.getAnsiColorCode(mappedColor, true);
        }
        
        output += message;
        
        // Reset colors
        if (foregroundColor != null || backgroundColor != null) {
            output += '\x1b[0m';
        }
        
        process.stdout.write(output);
    }
   
    private static _debug: boolean = false;
    private static _verbose: boolean = false;
    private static _quiet: boolean = false;

    private static _cchLastStatus: number = 0;

    private static _allLinesFromStdin: string[] | null = null;
    
    private static getAnsiColorCode(color: ConsoleColor, isBackground: boolean): string {
        const colorMap: { [key in ConsoleColor]: number } = {
            [ConsoleColor.Black]: 0,
            [ConsoleColor.DarkBlue]: 4,
            [ConsoleColor.DarkGreen]: 2,
            [ConsoleColor.DarkCyan]: 6,
            [ConsoleColor.DarkRed]: 1,
            [ConsoleColor.DarkMagenta]: 5,
            [ConsoleColor.DarkYellow]: 3,
            [ConsoleColor.Gray]: 7,
            [ConsoleColor.DarkGray]: 8,
            [ConsoleColor.Blue]: 12,
            [ConsoleColor.Green]: 10,
            [ConsoleColor.Cyan]: 14,
            [ConsoleColor.Red]: 9,
            [ConsoleColor.Magenta]: 13,
            [ConsoleColor.Yellow]: 11,
            [ConsoleColor.White]: 15
        };
        
        const colorCode = colorMap[color] || 7; // Default to gray
        const baseCode = isBackground ? 40 : 30;
        
        if (colorCode < 8) {
            return `\x1b[${baseCode + colorCode}m`;
        } else {
            return `\x1b[${baseCode + 60 + colorCode - 8}m`;
        }
    }
}